docs/src/conventions.md

```diff
@@ -0,0 +1,116 @@
+For the class of errors that Kani can verify, we should not produce a “No Error” result if there was in fact an
+error in the code being verified, i.e., it has no
+“False Negatives”.
+
+Because of that, we caution on the side of correctness.
+Any incorrect modeling
+that may trigger an unsound analysis that cannot be fixed in the short term should be mitigated.
+Here are a few ways how we do that.
+
+### Compilation errors
+
+Make sure to add user-friendly errors for constructs that we can't handle.
+For example, Kani cannot handle panic unwind strategy, and it will fail compilation if the crate uses this
+configuration.
+
+### Internal compiler errors
+
+Even though this doesn't provide users the best experience, you are encouraged to add checks in the compiler for any
+assumptions you make during development.
+Those check can be on the form of `assert!()` or `unreachable!()`
+statement.
+Please provide a meaningful message to help user understand why something failed, and also help us figure out what
+went wrong.
+
+We don't formally use any specific formal representation of [function contract](https://en.wikipedia.org/wiki/Design_by_contract),
+but whenever possible we do instrument the code with assertions that may represent the function pre- and
+post-conditions to ensure we are modeling the user code correctly.
+
+### Verification errors
+
+In cases where Kani fails to model a certain instruction or local construct that doesn't have a global effect,
+we encode this failure as a verification error.
+I.e., we generate an assertion failure instead of the construct we are modeling using
+[`codegen_unimplemented()`](https://github.com/model-checking/kani/blob/f719b565968568335d9be03ef27c5d05bb8fd0b7/kani-compiler/src/codegen_cprover_gotoc/utils/utils.rs#L50).
+
+This will allow users to verify their crate successfully as long as
+that construct is not reachable in any harness. If a harness has at least one possible execution path that reaches
+such construct, Kani will fail the verification and it will mark all checks with `UNDETERMINED` status.
+
+### Create detailed issues for "TODO" tasks
+
+It is OK to add "TODO" comments as long as they don't compromise user experience or the tool correctness.
+When doing so, please create an issue that captures the task.
+Add details about the task at hand including any impact to the user.
+Finally, add the link to the issue that captures the "TODO" task as part of your comment.
+
+E.g.:
+```rust
+// TODO: This function assumes type cannot be ZST. Check if that's always the case.
+// https://github.com/model-checking/kani/issues/XXXX
+assert!(!typ.is_zst(), "Unexpected ZST type");
+```
+
+## Performant but readable
+
+We aim at writing code that is performant but also readable and easy to maintain.
+Avoid compromising the code quality if the performance gain is not significant.
+
+Here are few tips that can help the readability of your code:
+
+- Sort match arms alphabetically.
```

Comment:
There are a number of places where we don't and shouldn't. I'd at least add a qualification "when possible" or something like that.

Reply:
Totally. There are places like inside a `map()` statement that it totally makes to just use something like `s` for string, but we do tend to over do it.

I almost want to add a rule here saying that it's not because it has been done before that it is a good idea. :)